### PR TITLE
Pass  "extra" substitutions from designspace rules to classifyGlyphs

### LINKED ISSUE #730: Alternate layers / FeatureVariations glyphs not taking part in mark attachment
I just noticed that although my base layers in a variable Oriya font were attaching nicely in a `blwm` mark-to-mark feature, alternate layers were not:

oriya

This turns out to be because we build a mark filtering class for mark features like so:

https://github.com/googlefonts/ufo2ft/blob/0f123a791c62ccd09ad41992f6cd52ed21858bc2/Lib/ufo2ft/featureWriters/markFeatureWriter.py#L678-L679

where include is something like this:

https://github.com/googlefonts/ufo2ft/blob/0f123a791c62ccd09ad41992f6cd52ed21858bc2/Lib/ufo2ft/featureWriters/markFeatureWriter.py#L828-L831

However:

```
In [10]: lookupName
Out[10]: 'blwm_mark2mark_bottom'

In [11]: include
Out[11]: <function ufo2ft.featureWriters.markFeatureWriter.MarkFeatureWriter._makeFeatures.<locals>.isAbvm(glyphName)>

In [12]: include("lVocalicMatra-oriya")
Out[12]: True

In [13]: include("lVocalicMatra-oriya.BRACKET.varAlt01")
Out[13]: False
```

`lVocalicMatra-oriya` is included in the glyph set because it has a Unicode encoded codepoint pointing to an abvm-using script; and anything reachable *by GSUB substitutions* from that codepoint is also included:

https://github.com/googlefonts/ufo2ft/blob/0f123a791c62ccd09ad41992f6cd52ed21858bc2/Lib/ufo2ft/featureWriters/markFeatureWriter.py#L878-L883

But (and here I'm guessing slightly but it's a reasonable guess) `lVocalicMatra-oriya.BRACKET.varAlt01` is only reachable by *feature variations*, so is not included, so does not take part in mark attachment.

I think the fix here is that either `compileGSUB` or `classifyGlyphs` or both need to know more about feature variations.

### PR DESCRIPTION
**Pull request by simoncozens**
Fixes #730

### CODE REVIEW

File: Lib/ufo2ft/featureCompiler.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ class BaseFeatureCompiler:
     layout tables from these.
     """
 
-    def __init__(self, ufo, ttFont=None, glyphSet=None, **kwargs):
```
Comment (anthrotype) [inline Lib/ufo2ft/featureCompiler.py:77]: I can't exactly remember, but maybe that `**kwargs` that you removed was intentional?

Comment (anthrotype) [reply Lib/ufo2ft/featureCompiler.py:77]: nevermind, it seems it's ok to remove it.

by the way, the BaseFeatureCompiler.__init__ has a docstring describing each Args, I think you could add an entry for the new extraSubstitutions there

File: tests/data/Alternates-Regular.ufo/fontinfo.plist [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    <key>capHeight</key>
+    <integer>714</integer>
+    <key>descender</key>
+    <integer>-293</integer>
+    <key>familyName</key>
+    <string>Alternates</string>
+    <key>guidelines</key>
+    <array>
+      <dict>
+        <key>angle</key>
```
Comment (anthrotype) [inline tests/data/Alternates-Regular.ufo/fontinfo.plist:16]: I wonder if you could trim unneeded pieces of data like the guidelines here, to make this test ufo smaller and more manageable

File: tests/data/Alternates-Regular.ufo/lib.plist [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+      <dict>
+        <key>name</key>
+        <string>eraseOpenCorners</string>
+        <key>namespace</key>
+        <string>glyphsLib.filters</string>
+        <key>pre</key>
+        <true/>
+      </dict>
+    </array>
+    <key>com.schriftgestaltung.Glyphs.glyphUserData.ka-oriya.below</key>
```
Comment (anthrotype) [inline tests/data/Alternates-Regular.ufo/lib.plist:16]: ... same with all the `com.schriftgestaltung.*` private glyphsLib stuff that ufo2ft doesn't care about

File: tests/featureWriters/markFeatureWriter_test.py [hunk 4]
```diff
@@ -1415,6 +1415,29 @@ def test_quantize(self, testufo):
...
+    def test_extra_substitutions(self, FontClass):
+        dirname = os.path.dirname(os.path.dirname(__file__))
+        fontPath = os.path.join(dirname, "data", "Alternates-Regular.ufo")
+        testufo = FontClass(fontPath)
+        generated = self.writeFeatures(
+            testufo,
+            compiler=FeatureCompiler(
+                testufo,
+                extraSubstitutions={
+                    "uuMatra-oriya": set(["uuMatra-oriya.BRACKET.varAlt01"]),
```
Comment (anthrotype) [inline tests/featureWriters/markFeatureWriter_test.py:1427]: you could use set literals `{"uuMatra-oriya.BRACKET.varAlt01"}` instead of `set([...])`

**Review by anthrotype — APPROVED** [review]

File: Lib/ufo2ft/util.py [hunk 5]
```diff
@@ -317,6 +320,13 @@ def classifyGlyphs(unicodeFunc, cmap, gsub=None):
             closeGlyphsOverGSUB(gsub, s)
             glyphs.update(s - neutralGlyphs)
 
+    if extra_substitutions:
+        for glyphs in glyphSets.values():
+            to_append = set()
+            for glyph in glyphs:
+                to_append |= extra_substitutions.get(glyph, set())
```
Comment (madig) [inline Lib/ufo2ft/util.py:327]: Shouldn't this happen before we do a GSUB closure because these substituted glyphs could be substituted further?

Comment (simoncozens) [reply Lib/ufo2ft/util.py:327]: Hrm, probably.

**Pull request merged**